internal/lib/checkpoint.go

```diff
@@ -54,8 +81,86 @@ func (c *ContainerServer) ContainerCheckpoint(ctx context.Context, opts *Contain
 		return "", fmt.Errorf("failed to checkpoint container %s: %w", ctr.ID(), err)
 	}
 	if opts.TargetFile != "" {
-		if err := c.exportCheckpoint(ctr, specgen.Config, opts.TargetFile); err != nil {
-			return "", fmt.Errorf("failed to write file system changes of container %s: %w", ctr.ID(), err)
+		logrus.Debugf("Trying to save the checkpoint as %s", opts.TargetFile)
+		// This check is a workaround for the limited possibilities presented
+		// by the current CRI CheckpointContainer API. The API does not allow
+		// to distinguish if an image should be a local tar archive on the
+		// file system or an OCI image in the local image store.
+		// This is again a feature brought over from Podman and especially
+		// useful if the user wants to start a container from a checkpoint
+		// image in a registry. Fast container startup from a checkpoint
+		// is one of the main use cases for this.
+		// The check here looks if the TargetFile parameter starts with a '/'
+		// or if it does not contain a slash. This means that the user wants
+		// to store in a local file. Either '/tmp/checkpoint-archive.tar' or
+		// 'checkpoint-archive.tar'. If the parameter TargetFile contains
+		// slashes in any other places than the front we assume it is something
+		// like 'localhost/checkpoint-archive:tag' and an OCI checkpoint image
+		// is created in the local image store with the help of buildah.
+		//
+		// As the crio.sock is only accessible by root and CRIU support currently
+		// needs to be enable explicitly in CRI-O only root is able to trigger
+		// a checkpoint request.
+		if strings.HasPrefix(opts.TargetFile, "/") || !strings.Contains(opts.TargetFile, "/") {
+			if err := c.exportCheckpoint(ctr, specgen.Config, opts.TargetFile); err != nil {
+				return "", fmt.Errorf("failed to write file system changes of container %s: %w", ctr.ID(), err)
+			}
+		} else {
+			// Create a checkpoint image in the local image store
+			imageRef, err := istorage.Transport.ParseStoreReference(c.store, opts.TargetFile)
```

Comment:
nit: can this block be a separate function?

Reply:
done